crates/ide/src/hover.rs

```diff
@@ -111,100 +113,179 @@ pub(crate) fn hover(
+                        // lints
+                        if let Some(res) = try_hover_for_lint(&attr, &token) {
+                            return Some(ControlFlow::Break(res));
+                        // derives
+                        } else {
+                            range_override = Some(token.text_range());
+                            try_resolve_derive_input_at(&sema, &attr, &token).map(Definition::Macro)
+                        }
                     } else {
-                        range_override = Some(token.text_range());
-                        try_resolve_derive_input_at(&sema, &attr, &token).map(Definition::Macro)
+                        None
                     }
-                } else {
-                    None
-                }
-            },
-        }
-    };
+                },
+            }
+        };
 
     if let Some(definition) = definition {
+        // skip duplicates
+        if seen.contains(&definition) {
+            return None;
+        }
+        if add_to_seen_definitions {
+            seen.insert(definition);
+        }
         let famous_defs = match &definition {
             Definition::ModuleDef(hir::ModuleDef::BuiltinType(_)) => {
                 Some(FamousDefs(&sema, sema.scope(&node).krate()))
             }
             _ => None,
         };
-        if let Some(markup) = hover_for_definition(db, definition, famous_defs.as_ref(), config) {
+        if let Some(markup) =
+            hover_for_definition(sema.db, definition, famous_defs.as_ref(), config)
+        {
             let mut res = HoverResult::default();
             res.markup = process_markup(sema.db, definition, &markup, config);
-            if let Some(action) = show_implementations_action(db, definition) {
+            if let Some(action) = show_implementations_action(sema.db, definition) {
                 res.actions.push(action);
             }
 
-            if let Some(action) = show_fn_references_action(db, definition) {
+            if let Some(action) = show_fn_references_action(sema.db, definition) {
                 res.actions.push(action);
             }
 
             if let Some(action) = runnable_action(&sema, definition, file_id) {
                 res.actions.push(action);
             }
 
-            if let Some(action) = goto_type_action_for_def(db, definition) {
+            if let Some(action) = goto_type_action_for_def(sema.db, definition) {
                 res.actions.push(action);
             }
 
             let range = range_override.unwrap_or_else(|| sema.original_range(&node).range);
-            return Some(RangeInfo::new(range, res));
+            return Some(ControlFlow::Break(RangeInfo::new(range, res)));
         }
     }
 
-    if let res @ Some(_) = hover_for_keyword(&sema, config, &token) {
-        return res;
+    if let Some(res) = hover_for_keyword(&sema, config, &token) {
+        return Some(ControlFlow::Break(res));
     }
```

Comment:
This should go into the fallback bag, we do not want to show keyword info if something also resolves to a non-keyword thing.

Reply:
ah, hm, then I can't use `ControlFlow` after all. So, back to `Either` - or should I write a custom enum with an explitly named `WithFallback(Option<...>)` variant?

Reply:
How so? You should be able to just move this out to the same place as `type_hover` no?

Reply:
misunderstanding on my side, sorry, will fix

Reply:
like so? dcc0ea7

Reply:
No I was thinking o something along the lines of
```rs
if fallback.is_none() {
    // FIXME we're only taking the first fallback into account that's not `None`
    fallback = type_hover(&sema, config, &token).or_else(|| hover_for_keyword(&sema, config, &token));
}
```
in https://github.com/spookyvision/rust-analyzer/blob/8e854de71bb4e5e27b914506baf704464048927b/crates/ide/src/hover.rs#L128-L131

Reply:
ok, so it's a fallback-fallback 😄